src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -415,6 +422,145 @@ private void HandleCodeActionRequest(string id, string paramsJson)
+
+            HandleSemanticTokens(id, semanticTokensParams, TextDocumentNames.RangeDocumentSemanticTokens);
+        }
+
+        private void HandleSemanticTokens<T>(string id, T semanticTokensParams, string method) 
+            where T : SemanticTokensParams
+        {
+            var uri = new Uri(semanticTokensParams.TextDocument.Uri);
+            var queryParams = HttpUtility.ParseQueryString(uri.Query);
+            var expression = queryParams?.Get("expression") ?? string.Empty;
+
+            if (string.IsNullOrEmpty(expression))
+            {
+                // Empty tokens for the empty expression
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var isRangeSemanticTokensMethod = method == TextDocumentNames.RangeDocumentSemanticTokens;
+
+            // Monaco-Editor sometimes uses \r\n for the newline character. \n is not always the eol character so allowing clients to pass eol character
+            var eol = queryParams?.Get("eol");
+            eol = !string.IsNullOrEmpty(eol) ? eol : EOL.ToString();
+            
+            var startIndex = -1;
+            var endIndex = -1;
+            if (isRangeSemanticTokensMethod)
+            {
+                (startIndex, endIndex) = PositionRangeHelper.ConvertRangeToPositions((semanticTokensParams as SemanticTokensRangeParams).Range, expression, eol);
+                if (startIndex < 0 || endIndex < 0)
+                {
+                    SendEmptySemanticTokensResponse(id);
+                    return;
+                }
+            }
+
+            var tokenTypesToSkip = ParseTokenTypesToSkipParam(queryParams?.Get("tokenTypesToSkip"));
+            var scope = _scopeFactory.GetOrCreateInstance(semanticTokensParams.TextDocument.Uri);
+            var result = scope.Check(expression);
+
+            // Skip over the token types that clients don't want in the response
+            var tokens = result.GetTokens().Where(tok => !tokenTypesToSkip.Contains(tok.TokenType));
+
+            if (isRangeSemanticTokensMethod)
+            {
+                // Only consider overlapping tokens. end index is exlcusive
+                tokens = tokens.Where(token => !(token.EndIndex <= startIndex || token.StartIndex >= endIndex));
+            }
+
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol);
+            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+        }
+
+        private static HashSet<TokenType> ParseTokenTypesToSkipParam(string rawTokenTypesToSkipParam)
+        {
+            var tokenTypesToSkip = new HashSet<TokenType>();    
+            if (string.IsNullOrWhiteSpace(rawTokenTypesToSkipParam))
+            {
+                return tokenTypesToSkip;
+            }
+
+            try
+            {
+                var tokenTypesToSkipParam = JsonSerializer.Deserialize<List<int>>(rawTokenTypesToSkipParam);
```

Comment:
> JsonSerializer.Deserialize<List<i [](http://example.com/codeflow?start=43&length=34)

use TryParseParams? and then avoid try/catch

Reply:
Refactored this to use TryParseParams